### PR TITLE
fix/TR-4934/session cookie and set cookie parameters in array

### PR DESCRIPTION
**Pull request by africz**
# [TR-4934](https://oat-sa.atlassian.net/browse/TR-4934)
**Description**
session_set_cookie_params and setcookie calls adjusted to latest calling form available from php7.3

https://www.php.net/manual/en/function.setcookie.php
https://www.php.net/manual/en/function.session-set-cookie-params.php

session_set_cookie_params(array $lifetime_or_options): bool
setcookie(string $name, string $value = "", array $options = []): bool

**Development impact**
1. I did apply at initialization the default options such as lifetime, domain, secure, httponly. In reset cookie session lifetime replaced to expires. 
2. I did test modification with nextgen backoffice and unit tests, both works fine, either with all domain allowed or restricted to current one.

### CODE REVIEW

**Comment by africz** [comment]
I take a second attempt now, to move the init function if possible to factory interface and make them available before create function is executed.

File: models/classes/session/Business/Contract/SessionCookieAttributeInterface.php [hunk 1]
```diff
@@ -27,4 +27,9 @@
 interface SessionCookieAttributeInterface
 {
     public function __toString(): string;
+
+    public function getName(): string;
+
+    public function getValue();
+
```
Comment (poyuki) [inline models/classes/session/Business/Contract/SessionCookieAttributeInterface.php:34]:
```suggestion
```

File: models/classes/session/Business/Domain/SessionCookieAttribute.php [hunk 2]
```diff
@@ -43,4 +44,16 @@ public function __toString(): string
...
+    {
+        return $this->name;
+    }
+
+
+    public function getValue()
+    {
+        return $this->value;
+    }
+
```
Comment (poyuki) [inline models/classes/session/Business/Domain/SessionCookieAttribute.php:58]:
```suggestion

    public function getName(): string
    {
        return $this->name;
    }

    public function getValue()
    {
        return $this->value;
    }
```

File: models/classes/session/Business/Domain/SessionCookiePathAttribute.php [hunk 3]
```diff
@@ -36,6 +36,18 @@ public function __construct(string $value)
...
+        return 'path';
+    }
+
+
+    public function getValue()
+    {
+        return $this->value;
+    }
+
+
```
Comment (poyuki) [inline models/classes/session/Business/Domain/SessionCookiePathAttribute.php:50]:
```suggestion
    public function getName(): string
    {
        return 'path';
    }

    public function getValue()
    {
        return $this->value;
    }

```

File: models/classes/session/Business/Service/SessionCookieService.php [hunk 4]
```diff
@@ -15,67 +15,103 @@
...
  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  *
- * Copyright (c) 2020 (original work) Open Assessment Technologies SA;
+ * Copyright (c) 2020-2022 (original work) Open Assessment Technologies SA;
  *
  * @author Sergei Mikhailov <sergei.mikhailov@taotesting.com>
  */
 
-declare(strict_types=1);
+declare (strict_types=1);
```
Comment (poyuki) [inline models/classes/session/Business/Service/SessionCookieService.php:23]:
```suggestion
declare(strict_types=1);
```

Comment (poyuki) [inline models/classes/session/Business/Service/SessionCookieService.php:23]:
```suggestion
declare(strict_types=1);
```

File: models/classes/session/Business/Service/SessionCookieService.php [hunk 5]
```diff
@@ -15,67 +15,103 @@
...
+    private $attributes = [];
+    private $sessionCookieParams = [];
 
     public function __construct(SessionCookieAttributesFactoryInterface $sessionCookieAttributesFactory)
     {
         parent::__construct();
-
         $this->sessionCookieAttributesFactory = $sessionCookieAttributesFactory;
+        //it will brake the unit test logic
+        //   $this->init();
```
Comment (poyuki) [inline models/classes/session/Business/Service/SessionCookieService.php:46]: let's remove debug info
```suggestion
```

File: models/classes/session/Business/Service/SessionCookieService.php [hunk 6]
```diff
@@ -15,67 +15,103 @@
...
 use oat\tao\model\session\Business\Contract\SessionCookieAttributesFactoryInterface;
 use oat\tao\model\session\Business\Contract\SessionCookieServiceInterface;
+use oat\tao\model\session\Business\Domain\SessionCookieAttribute;
 use tao_helpers_Uri as UriHelper;
 
 class SessionCookieService extends InjectionAwareService implements SessionCookieServiceInterface
 {
     /** @var SessionCookieAttributesFactoryInterface */
     private $sessionCookieAttributesFactory;
+    private $attributes = [];
```
Comment (poyuki) [inline models/classes/session/Business/Service/SessionCookieService.php:38]: property attribute used only in context of `init` method. i don't see any needs to define it as property, lets just provide local variable inestead

File: models/classes/session/Business/Service/SessionCookieService.php [hunk 7]
```diff
@@ -15,67 +15,103 @@
...
     }
 
-    public function initializeSessionCookie(): void
+    private function init(): void
     {
+        $iterator = $this->sessionCookieAttributesFactory->create()->getIterator();
+        while ($iterator->valid()) {
+            $this->attributes[] = $iterator->current();
+            $iterator->next();
+        }
```
Comment (poyuki) [inline models/classes/session/Business/Service/SessionCookieService.php:55]:
```suggestion
        $attributes = [];
        $iterator = $this->sessionCookieAttributesFactory->create()->getIterator();
        while ($iterator->valid()) {
            $attributes[] = $iterator->current();
            $iterator->next();
        }
```
reason of it mentioned here https://github.com/oat-sa/tao-core/pull/3712/files#r1057772958

Comment (poyuki) [inline models/classes/session/Business/Service/SessionCookieService.php:55]:
```suggestion
        $attributes = [];
        $iterator = $this->sessionCookieAttributesFactory->create()->getIterator();
        while ($iterator->valid()) {
            $attributes[] = $iterator->current();
            $iterator->next();
        }
```
reason of it mentioned here https://github.com/oat-sa/tao-core/pull/3712/files#r1057772958

File: models/classes/session/Business/Service/SessionCookieService.php [hunk 8]
```diff
@@ -15,67 +15,103 @@
...
-        $isSecureFlag  = Request::isHttps();
-        $sessionCookieAttributes = $this->sessionCookieAttributesFactory->create();
+        $isSecureFlag = Request::isHttps();
+        if (isset($sessionParams['lifetime'])) {
+            $this->attributes[] = new SessionCookieAttribute('lifetime', $sessionParams['lifetime']);
+        }
+
+        $this->attributes[] = new SessionCookieAttribute('domain', $cookieDomain);
+        $this->attributes[] = new SessionCookieAttribute('secure', $isSecureFlag);
+        $this->attributes[] = new SessionCookieAttribute('httponly', true);
```
Comment (poyuki) [inline models/classes/session/Business/Service/SessionCookieService.php:68]:
```suggestion
        if (isset($sessionParams['lifetime'])) {
            $attributes[] = new SessionCookieAttribute('lifetime', $sessionParams['lifetime']);
        }

        $attributes[] = new SessionCookieAttribute('domain', $cookieDomain);
        $attributes[] = new SessionCookieAttribute('secure', $isSecureFlag);
        $attributes[] = new SessionCookieAttribute('httponly', true);
```
reason of that mentioned https://github.com/oat-sa/tao-core/pull/3712/files#r1057772958

File: models/classes/session/Business/Service/SessionCookieService.php [hunk 9]
```diff
@@ -15,67 +15,103 @@
...
+
+        foreach ($this->attributes as $attribute) {
```
Comment (poyuki) [inline models/classes/session/Business/Service/SessionCookieService.php:70]:
```suggestion
        foreach ($attributes as $attribute) {
```
reason of suggestion mentioned here https://github.com/oat-sa/tao-core/pull/3712/files#r1057772958

File: models/classes/session/Business/Service/SessionCookieService.php [hunk 10]
```diff
@@ -15,67 +15,103 @@
...
+        return $cookieParams;
+    }
+
+    public function initializeSessionCookie(): void
+    {
+        //call it here not in the constructor, to keep original logic
+        //of unit test
+        $this->init();
+        session_set_cookie_params($this->getSessionCookieParams());
+        //temporary line to verify replace is working $tmp = $this->setExpires($params);
```
Comment (poyuki) [inline models/classes/session/Business/Service/SessionCookieService.php:102]:
```suggestion
```

File: models/classes/session/Business/Contract/SessionCookieAttributeInterface.php [hunk 11]
```diff
@@ -27,4 +27,9 @@
 interface SessionCookieAttributeInterface
 {
     public function __toString(): string;
+
+    public function getName(): string;
+
+    public function getValue();
```
Comment (poyuki) [inline models/classes/session/Business/Contract/SessionCookieAttributeInterface.php:33]: extending of public interface actually is breaking change

to enforce bumping of major version  by release action
you can provide commit like `feat!: extending public contract of `SessionCookieAttributeInterface` `

Comment (africz) [reply models/classes/session/Business/Contract/SessionCookieAttributeInterface.php:33]: Not necessary to make this mod, works without this one as well if I not modify the interface.

Comment (poyuki) [reply models/classes/session/Business/Contract/SessionCookieAttributeInterface.php:33]: definitely it works but in collection we rely on interface but in common in this collection we rely on this interface https://github.com/oat-sa/tao-core/blob/6c1c3c6b35bbe2ba57d2ad48fe9f43b4157d3982/models/classes/session/Business/Domain/SessionCookieAttributeCollection.php#L30

**Review by poyuki — APPROVED** [review]
- [ ] New code is covered by tests (if applicable)
- [x] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [x] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful
- [x] Pull request's target is not `master`
- [x] Commits are following [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [ ] Changelog is updated according to changes (if applicable)
- [ ] Documentation is updated according to changes (if applicable)

**Pull request merged**